systems/unix/system.go

```diff
@@ -700,6 +702,125 @@ func (s *System) SockRemoteAddress(ctx context.Context, fd wasi.FD) (wasi.Socket
 	return addr, wasi.ESUCCESS
 }
 
+func (s *System) SockAddressInfo(ctx context.Context, name, service string, hint wasi.AddressInfo, results []wasi.AddressInfo) (int, wasi.Errno) {
+	if cap(results) == 0 {
+		return 0, wasi.EINVAL
+	}
+	// TODO: support AI_ADDRCONFIG, AI_CANONNAME, AI_V4MAPPED, AI_V4MAPPED_CFG, AI_ALL
+
+	var network string
+	f, p, t := hint.Family, hint.Protocol, hint.SocketType
+	switch {
+	case t == wasi.StreamSocket && p != wasi.UDPProtocol:
+		switch f {
+		case wasi.UnspecifiedFamily:
+			network = "tcp"
+		case wasi.InetFamily:
+			network = "tcp4"
+		case wasi.Inet6Family:
+			network = "tcp6"
+		default:
+			return 0, wasi.ENOTSUP // EAI_FAMILY
+		}
+	case t == wasi.DatagramSocket && p != wasi.TCPProtocol:
+		switch f {
+		case wasi.UnspecifiedFamily:
+			network = "udp"
+		case wasi.InetFamily:
+			network = "udp4"
+		case wasi.Inet6Family:
+			network = "udp6"
+		default:
+			return 0, wasi.ENOTSUP // EAI_FAMILY
+		}
+	case t == wasi.AnySocket:
+		switch f {
+		case wasi.UnspecifiedFamily:
+			network = "ip"
+		case wasi.InetFamily:
+			network = "ip4"
+		case wasi.Inet6Family:
+			network = "ip6"
+		default:
+			return 0, wasi.ENOTSUP // EAI_FAMILY
+		}
+	default:
+		return 0, wasi.ENOTSUP // EAI_SOCKTYPE / EAI_PROTOCOL
+	}
+
+	var port int
+	var err error
+	if hint.Flags.Has(wasi.NumericService) {
+		port, err = strconv.Atoi(service)
+	} else {
+		port, err = net.LookupPort(network, service)
+	}
+	if err != nil || port < 0 || port > 65535 {
+		return 0, wasi.EINVAL // EAI_NONAME / EAI_SERVICE
+	}
+
+	var ip net.IP
+	if hint.Flags.Has(wasi.NumericHost) {
+		ip = net.ParseIP(name)
+		if ip == nil {
+			return 0, wasi.EINVAL
+		}
+	} else if name == "" {
+		if !hint.Flags.Has(wasi.Passive) {
+			return 0, wasi.EINVAL
+		}
+		if hint.Family == wasi.Inet6Family {
+			ip = net.IPv6zero
+		} else {
+			ip = net.IPv4zero
+		}
+	}
+	if ip != nil {
+		results = results[:1]
+		results[0] = wasi.AddressInfo{}
+		if ipv4 := ip.To4(); ipv4 != nil {
+			s.inet4Addr.Port = port
+			copy(s.inet4Addr.Addr[:], ipv4)
+			results[0].Address = &s.inet4Addr
+		} else {
+			s.inet6Addr.Port = port
+			copy(s.inet6Addr.Addr[:], ip)
+			results[0].Address = &s.inet6Addr
+		}
+		return 1, wasi.ESUCCESS
+	}
+
+	ips, err := net.LookupIP(name)
+	if err != nil {
+		return 0, wasi.ECANCELED // TODO: better errors on name resolution failure
```

Comment:
Any reason not to return `makeErrno(err)` here?

Reply:
The errors that it returns don't implement `syscall.Errno`, and aren't errors that we're converting at the moment. If we use `makeErrno` it'll almost certainly hit the panic [here](https://github.com/stealthrocket/wasi-go/blob/7a8e15e85f37f38cab1fb7f126ea46a62bb5772e/systems/unix/syscall_unix.go#L41). We need a translation layer. I'll look at doing this in a follow-up.